File: tree/traversal.cpp
#include<iostream>
using namespace std;

struct BSTNode
{
    int data;
    BSTNode* left;
    BSTNode* right;
};

BSTNode* root;

BSTNode* GetNewNode(int data)
{   
    BSTNode *temp = new BSTNode;  
    temp->data = data;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
};


BSTNode* insert(BSTNode* root, int data)
{
    if (root==NULL)
    {
        root = GetNewNode(data);
        return root;
    }

    if (data<=root->data) 
    {
        root->left = insert(root->left, data);
    }
    else
    {
        root->right = insert(root->right, data);
    }

    return root;
}

void inorder(BSTNode* root)
{
    if (root==NULL)
    {
        return ;
    }
        inorder(root->left);
        cout<<"data"<<root->data<<endl;
        inorder(root->right);
}

void preorder(BSTNode* root)
{
    if (root==NULL)
    {
        return;
    }
    cout<<"Data"<<root->data<<endl;
    preorder(root->left);
    preorder(root->right)
}

void postorder(BSTNode* root)
{
    if (root==NULL)
    {
        return;
    }
    postorder(root->left);
    postorder(root->right)
    cout<<"Data"<<root->data<<endl;
}



int main()
{
    root=NULL;
    cout<<"Implement Binary Tree"<<endl;
    int n;
    cout<<"Enter n value to be enter in tree ";
    cin>>n;
    for (int i = 0; i < n; i++)
    {
        int data;
        cout<<"Enter data in new node ";
        cin>>data;
        root = insert(root, data);
        cout<<endl;
        cout<<"root data "<<root->data<<endl;
        cout<<endl;
    }

    inorder(root);
    preorder(root);
    postorder(root);
    return 0;
}

File: tree/others/traversal.py
class Node:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.key = val

def insert(root, val):
    if root==None:
        root = Node(val)
        return root

    if val < root.key:
        root.left = insert(root.left, val)
    else:
        root.right = insert(root.right, val)
    
    return root

def inorder(root):
    if root is None:
        return

    inorder(root.left)
    print("data",root.key)
    inorder(root.right)

def preorder(root):
    if root is None:
        return

    print("data",root.key)
    preorder(root.left)
    preorder(root.right)

def postorder(root):
    if root is None:
        return
    postorder(root.left)
    postorder(root.right)
    print("data",root.key)


if __name__=='__main__':
    root = None
    print("Enter n value in BST")
    n = int(input())
    for i in range(n):
        val = int(input())
        root = insert(root, val)
    
    print("Inorder")
    inorder(root)
    print("Preorder")
    preorder(root)
    print("Postorder")
    postorder(root)
